src/Modules/UngradedModules.jl

```diff
@@ -6966,3 +6988,73 @@ function change_base_ring(f::Hecke.Map{DomType, CodType}, M::SubQuo) where {DomT
   map = SubQuoHom(M, MS, gens(MS), f)
   return MS, map
 end
+
+### Duals of modules
+@Markdown.doc """
+    dual(M::ModuleFP; cod::FreeMod=FreeMod(base_ring(M), 1))
+
+Return a pair ``(M*, i)`` consisting of the dual of ``M`` and its 
+interpretation map ``i``, turning an element ``φ`` of ``M*`` into 
+a homomorphism ``M → R``. 
+
+The optional argument allows to specify a free module of rank ``1`` 
+for the codomain of the dualizing functor.
+"""
+function dual(M::ModuleFP; cod::FreeMod=FreeMod(base_ring(M), 1))
+  base_ring(cod) === base_ring(M) && rank(cod) == 1 || error("codomain must be free of rank one over the base ring of the first argument")
+  return hom(M, cod)
+end
+
+@Markdown.doc """
+    double_dual(M::ModuleFP)
+
+For a finite ``R``-module ``M`` return a pair ``(M**, ϕ)`` consisting of 
+its double dual ``M** = Hom(Hom(M, R), R)`` together with the canonical 
+map ``ϕ : M → M**, v ↦ (φ ↦ φ(v)) ∈ Hom(M*, R)``.
+"""
+function double_dual(M::ModuleFP; cod::FreeMod=FreeMod(base_ring(M), 1))
+  M_dual, _ = dual(M, cod=cod)
+  M_double_dual, _ = dual(M_dual, cod=cod)
+  psi = hom(M, M_double_dual, 
+            [homomorphism_to_element(M_double_dual, 
+                                     hom(M_dual, cod,
+                                         [element_to_homomorphism(phi)(x) for phi in gens(M_dual)]
+                                        )
+                                    )
+             for x in gens(M)
+            ]
+           )
+  return M_double_dual, psi
+end
+
+@Markdown.doc """
+    dual(f::ModuleFPHom; cod::FreeMod)
+
+For a morphism of modules ``f : M → N`` this returns the morphism 
+``fᵀ : N* → M*, φ ↦ (v ↦ φ(f(v)))`` induced on the duals.
+
+The optional argument allows to specify a free module over the 
+base ring of ``f`` for building the duals of ``M`` and ``N``.
```

Comment:
No clue what this means mathematically. For the duals I know it is always `hom_R(M,R)` where `M` is an `R`-module.

Reply:
Yes. But on the computer side, this is different. `R` is not just `R`, but one specific free module of rank 1, located somewhere in memory. `dual` returns a map which allows to evaluate stuff and that lands in that very module. So you might want to specify, where you land.